File: Java/sorting/SelectionSort.java
/**
 * Selection Sort Algorithm
 * 
 * Time Complexity: O(n^2) in all cases (worst, average, and best)
 * Space Complexity: O(1)
 * 
 * Selection sort is a simple sorting algorithm that divides the input list into
 * two parts:
 * a sorted portion at the left end and an unsorted portion at the right end.
 * Initially, the sorted portion is empty and the unsorted portion is the entire
 * list.
 * The algorithm proceeds by finding the smallest (or largest) element in the
 * unsorted portion,
 * swapping it with the leftmost unsorted element, and moving the sublist
 * boundaries one element to the right.
 */
public class SelectionSort {

  /**
   * Sorts an array using selection sort algorithm (ascending order).
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array
   */
  public static int[] selectionSort(int[] arr) {
    int n = arr.length;

    // Traverse through all array elements
    for (int i = 0; i < n - 1; i++) {
      // Find the minimum element in unsorted portion
      int minIndex = i;

      for (int j = i + 1; j < n; j++) {
        if (arr[j] < arr[minIndex]) {
          minIndex = j;
        }
      }

      // Swap the found minimum element with the first element
      if (minIndex != i) {
        int temp = arr[minIndex];
        arr[minIndex] = arr[i];
        arr[i] = temp;
      }
    }

    return arr;
  }

  /**
   * Sorts an array in descending order using selection sort.
   * 
   * @param arr Array of integers to be sorted
   * @return Sorted array in descending order
   */
  public static int[] selectionSortDescending(int[] arr) {
    int n = arr.length;

    for (int i = 0; i < n - 1; i++) {
      // Find the maximum element in unsorted portion
      int maxIndex = i;

      for (int j = i + 1; j < n; j++) {
        if (arr[j] > arr[maxIndex]) {
          maxIndex = j;
        }
      }

      // Swap the found maximum element with the first element
      if (maxIndex != i) {
        int temp = arr[maxIndex];
        arr[maxIndex] = arr[i];
        arr[i] = temp;
      }
    }

    return arr;
  }

  /**
   * Utility method to print an array
   */
  public static void printArray(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  /**
   * Main method with example usage
   */
  public static void main(String[] args) {
    System.out.println("Selection Sort Examples:");

    // Test ascending sort
    int[] arr1 = { 64, 34, 25, 12, 22, 11, 90 };
    System.out.print("\nOriginal array: ");
    printArray(arr1);

    selectionSort(arr1);
    System.out.print("Sorted array (ascending): ");
    printArray(arr1);

    // Test descending sort
    int[] arr2 = { 64, 34, 25, 12, 22, 11, 90 };
    selectionSortDescending(arr2);
    System.out.print("Sorted array (descending): ");
    printArray(arr2);

    // Test with already sorted array
    int[] arr3 = { 1, 2, 3, 4, 5 };
    System.out.print("\nAlready sorted array: ");
    printArray(arr3);
    selectionSort(arr3);
    System.out.print("After selection sort: ");
    printArray(arr3);

    // Test with single element
    int[] arr4 = { 42 };
    System.out.print("\nSingle element array: ");
    printArray(arr4);
    selectionSort(arr4);
    System.out.print("After selection sort: ");
    printArray(arr4);
  }
}

File: Python/sorting/selection_sort.py
"""
Selection Sort Algorithm
"""


def selection_sort(arr: list) -> list:
    """
    Selection Sort Algorithm
    
    Divides the array into sorted and unsorted portions. Repeatedly finds
    the minimum element from the unsorted portion and places it at the end
    of the sorted portion.
    
    Time Complexity: O(n²) - always, regardless of input
    Space Complexity: O(1) - sorts in place, only constant extra space
    
    Args:
        arr: List of comparable elements to be sorted
    
    Returns:
        Sorted list in ascending order
    
    Example:
        >>> selection_sort([64, 34, 25, 12, 22, 11, 90])
        [11, 12, 22, 25, 34, 64, 90]
    """
    n = len(arr)
    
    # Traverse through all array elements
    for i in range(n):
        # Find the minimum element in remaining unsorted array
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        
        # Swap the found minimum element with the first element
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    
    return arr


if __name__ == "__main__":
    # Test Case 1
    arr1 = [64, 34, 25, 12, 22, 11, 90]
    print(f"Original array: {arr1}")
    result1 = selection_sort(arr1.copy())
    print(f"Sorted array: {result1}")
    
    # Test Case 2
    arr2 = [5, 2, 8, 1, 9]
    print(f"\nOriginal array: {arr2}")
    result2 = selection_sort(arr2.copy())
    print(f"Sorted array: {result2}")
    
    # Test Case 3
    arr3 = [1]
    print(f"\nOriginal array: {arr3}")
    result3 = selection_sort(arr3.copy())
    print(f"Sorted array: {result3}")
